internal/api/v1/deploy/deploy.go

```diff
@@ -0,0 +1,192 @@
+package deploy
+
+import (
+	"context"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/epinio/epinio/helpers/kubernetes"
+	"github.com/epinio/epinio/internal/application"
+	"github.com/epinio/epinio/internal/cli/server/requestctx"
+	"github.com/epinio/epinio/internal/helm"
+	"github.com/epinio/epinio/internal/helmchart"
+	"github.com/epinio/epinio/internal/registry"
+	apierror "github.com/epinio/epinio/pkg/api/core/v1/errors"
+	"github.com/epinio/epinio/pkg/api/core/v1/models"
+)
+
+func DeployCurrentAppState(ctx context.Context, cluster *kubernetes.Cluster, app models.AppRef, username, imageURL, stageID string, origin *models.ApplicationOrigin) ([]string, apierror.APIErrors) {
+	log := requestctx.Logger(ctx)
+
+	// check application resource
+	applicationCR, err := application.Get(ctx, cluster, app)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, apierror.AppIsNotKnown("cannot deploy app, application resource is missing")
+		}
+		return nil, apierror.InternalError(err, "failed to get the application resource")
+	}
+
+	if stageID == "<APP>" {
+		stageID, err = application.StageID(applicationCR)
+		if err != nil {
+			return nil, apierror.InternalError(err, "failed to retrieve application's stage id")
+		}
+	}
+	if imageURL == "" {
+		// Updating endpoint (update, env set, env unset) - Read image url from application.
+		imageURL, err = application.ImageURL(applicationCR)
+		if err != nil {
+			return nil, apierror.InternalError(err, "failed to retrieve application's image url")
+		}
+	} else {
+		// Deployment endpoint - Save image url to application for the updating endpoints.
+		err := updateImageURL(ctx, cluster, applicationCR, imageURL)
+		if err != nil {
+			return nil, apierror.InternalError(err, "failed to set application's image url")
+		}
+	}
+
+	// owner data to set into new resources
+	owner := metav1.OwnerReference{
+		APIVersion: applicationCR.GetAPIVersion(),
+		Kind:       applicationCR.GetKind(),
+		Name:       applicationCR.GetName(),
+		UID:        applicationCR.GetUID(),
+	}
+
+	// determine number of desired instances
+	instances, err := application.Scaling(ctx, cluster, app)
+	if err != nil {
+		return nil, apierror.InternalError(err, "failed to access application's desired instances")
+	}
+
+	// determine runtime environment, if any
+	// TODO: Used in helm.Deploy with environment.List().Assignments() - Simplify
```

Comment:
No, I do not believe that this TODO is needed anymore. Not as an issue either.

My thinking has changed.
I was not satisfied with having a chain of transforms for the environment
(map(name->value) -> slice (pair) -> slice (yaml-string)).
It is growing on me however, as clearer to read/understand.
Getting the data ordered and converted to yaml are kept separate, instead of mingled.

Action: Removing the comment.

Reply:
Done, commit e713fc21cfce6b2f7905a1577a64640418d41899